workflow/src/main/kotlin/com/r3/corda/sdk/token/workflow/selection/VaultWatcherService.kt

```diff
@@ -0,0 +1,143 @@
+package com.r3.corda.sdk.token.workflow.selection
+
+import com.r3.corda.sdk.token.contracts.states.FungibleToken
+import com.r3.corda.sdk.token.contracts.types.IssuedTokenType
+import com.r3.corda.sdk.token.contracts.types.TokenType
+import net.corda.core.contracts.Amount
+import net.corda.core.contracts.StateAndRef
+import net.corda.core.node.AppServiceHub
+import net.corda.core.node.services.CordaService
+import net.corda.core.node.services.Vault
+import net.corda.core.node.services.vault.DEFAULT_PAGE_NUM
+import net.corda.core.node.services.vault.PageSpecification
+import net.corda.core.utilities.contextLogger
+import rx.Observable
+import java.lang.RuntimeException
+import java.security.PublicKey
+import java.time.Duration
+import java.util.concurrent.*
+
+val unlocker: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
+
+@CordaService
+class VaultWatcherService(vaultObserver: VaultObserver? = null) {
+
+    val cache: ConcurrentMap<TokenIndex, TokenBucket> = ConcurrentHashMap()
+
+    constructor(appServiceHub: AppServiceHub) : this(getObservableFromAppServiceHub(appServiceHub))
+
+    companion object {
+        val LOG = contextLogger()
+
+        private fun getObservableFromAppServiceHub(appServiceHub: AppServiceHub): VaultObserver {
+            val pageSize = 1000
+            var currentPage = DEFAULT_PAGE_NUM;
+            var (existingStates, observable) = appServiceHub.vaultService.trackBy(FungibleToken::class.java, PageSpecification(pageNumber = currentPage, pageSize = pageSize))
+            val listOfThings = mutableListOf<StateAndRef<FungibleToken<TokenType>>>()
+            while (existingStates.states.isNotEmpty()) {
+                listOfThings.addAll(existingStates.states as Iterable<StateAndRef<FungibleToken<TokenType>>>)
+                existingStates = appServiceHub.vaultService.queryBy(FungibleToken::class.java, PageSpecification(pageNumber = ++currentPage, pageSize = pageSize))
+            }
+            return VaultObserver(listOfThings, observable)
+        }
+
+        fun processToken(token: FungibleToken<*>): TokenIndex {
+            val owner = token.holder.owningKey
+            val type = token.amount.token.tokenType.tokenClass
+            val typeId = token.amount.token.tokenType.tokenIdentifier
+            return TokenIndex(owner, type, typeId)
```

Comment:
How come you are indexing by owning key? This could potentially be different for each token, right? So you'd end up with a key in the cache for _each_ token? I get why you'd want to do it by x500 name though. Presumably you just need to index by token type (class and ID) to get a bucket per token type. Then if we want to index by, say... notary, issuer, or whatever, then we can add other index types?